contract/signals.py

```diff
@@ -58,6 +63,47 @@ def on_contract_approve_signal(sender, **kwargs):
 signal_contract_approve.connect(on_contract_approve_signal, dispatch_uid="on_contract_approve_signal")
 
 
+@receiver(post_save, sender=Payment, dispatch_uid="payment_signal_paid")
+def activate_contracted_policies(sender, instance, **kwargs):
+    received_amount = instance.received_amount if instance.received_amount else 0
+    # check if payment is related to the contract
+    uuid_payment = instance.uuid
+    if isinstance(uuid_payment, str):
+        uuid_payment = str(uuid.UUID(uuid_payment))
+    payment_detail = list(PaymentDetail.objects.filter(payment__id=int(instance.id)))
+    contract = None
+    if len(payment_detail) > 0:
+        # for  fetching contract getting the first payment
+        # detail from list would be enough to fetch it
+        contract = ContractContributionPlanDetails.objects.filter(contribution_id=payment_detail[0].premium.id).first().contract_details.contract
```

Comment:
A payment might concerns multiple contract, 
why not having related_name "contract_contribution_plan_details" for the contribution in the CCPD (https://github.com/openimis/openimis-be-contract_py/blob/05c1c5648ffbcca52a60450ca3b47db876249110/contract/models.py#L119) and having a query like
0- to check that this payment concerns a contract
PaymentDetail.filter(payment__id=int(instance.id)).select_related('premium__contract_contribution_plan_details__contract_details__contract).select_related('premium__contract_contribution_plan_details).filter(premium__contract_contribution_plan_details__isnull=false)

You could add the "only" statement to have only the fields you need (saves RAM)
 then if the payment include a contract payment:
1- check if the contract have payment attached to each contributions
(Nice to have)2- get all the payment related to all contribution that are part of the payment involving the contract (current and past)
3- check that  (nice to have: the sum of all) the payment match the value of all contributions
4-activate the insureepolicies of the contracts (can be severals)

if you don't do the nice to have please add todos